z80.h

```diff
@@ -2169,11 +2199,20 @@ class i8080_state : public internals::cpu_state_base<B> {
 public:
     typedef internals::cpu_state_base<B> base;
 
-    bool get_iff() const { return iff.get(); }
-    void set_iff(bool f) { iff.set(f); }
+    bool get_iff() const { return fields.iff.get(); }
+    void set_iff(bool f) { fields.iff.set(f); }
+
+    void on_reset_cpu() {
+        base::on_reset_cpu();
+        fields = state_fields();
```

Comment:
Won't this reset too many registers? I modified the unit test to set bc/de/hl/sp to known values, and they're gone after a reset, which doesn't match the real chip behaviour.

The Z80 has a power-on state with some registers holding known values. When the reset line is pulled low it's a relatively light reset, and different from the power-on reset state. Most registers are untouched, but PC, IR, IFF1 are reset (according to the official documentation). There may be some other internal state, such as the index prefix and halted that may need to be reset too.

Sean Young's _The Undocumented Z80 Documented_ says this about the power-on state:

> AF and SP are always set to FFFFh after a reset, and all other registers are undefined (different depending on how long the CPU has been powered off, different for different Z80 chips). Of course the PC should be set to 0 after a reset, and so should the IFF1 and IFF2 flags (otherwise strange things could happen). Also since the Z80 is 8080 compatible, interrupt mode is probably 0. Probably the best way to simulate this in an emulator is set PC, IFF1, IFF2, IM to 0 and set all other registers to FFFFh.

and this about reset state:
> it resets the interrupt flip-flops, clears the PC and IR registes, and set the interrupt mode to 0. During reset time, the
address bus and data bus go to a high-impedance state, and all control output signals go to the inactive state. Note that RESET must be active for a minimum of three full clock cycles before the reset operation is complete. Note that SP and AF are set to FFFFh.

Reply:
Thanks. Updated to support soft reset as well. The issue with initial values of `SP` and `AF` went to #31.